server/src/tracing.ts

```diff
@@ -0,0 +1,202 @@
+/**
+ * Tools for instrumenting code with traces. This is mainly designed to support
+ * Sentry's tracing features, but could be expanded support other approaches,
+ * like OpenTelemetry.
+ */
+
+import EventEmitter from "node:events";
+import * as Sentry from "@sentry/node";
+import { Transaction } from "@sentry/tracing";
+import type { Span, SpanStatusType } from "@sentry/tracing";
+import type { SpanContext } from "@sentry/types";
+export type { Span } from "@sentry/tracing";
+
+interface SpanOptions extends SpanContext {
+  parentSpan?: Span;
+  timeout?: number;
+}
+
+// Monkey-patch Transaction to add an event to notify listeners (in our case,
+// child spans) when the transaction is finishing. This is super hacky.
+//
+// The most recent release of the Sentry SDK has a `finishTransaction` event
+// on the hub's client, but there are a lot of guard clauses anywhere the client
+// gets used internally, and I'm not sure how reliable it is for this use case.
+// We want to wrap up before the transaction finishes, not when a client that
+// may-or-may-not exist depending on configuring is finishing a transaction.
+// (Also, the transaction's `_hub` is a nullable private property, so it would
+// still be hacky to grab it and add a listener anyway.)
+interface PatchedTransaction extends Transaction {
+  _onFinish(listener: (...args: any[]) => void): void;
+  _emitter?: EventEmitter;
+}
+
+const _transactionFinish = Transaction.prototype.finish;
+Transaction.prototype.finish = function finish(
+  this: PatchedTransaction,
+  endTimestamp?: number
+) {
+  this._emitter?.emit("finish", this);
+  return _transactionFinish.call(this, endTimestamp);
+};
+
+// @ts-expect-error: _onFinish doesn't exist; we're adding it.
+Transaction.prototype._onFinish = function onFinish(
```

Comment:
Any reason not to make this `onFinish` instead (without the underscore)?

Reply:
Mostly I was worried about colliding with some future feature. Then again, maybe it’s fine — if this method ever collided with an official feature, I’d be pretty darn surprised if was functionally different!